### ISSUE TITLE
Extensibility

### ISSUE THREAD

**Issue by SeeThruHead** [saguijs/sagui#166]
I've been hacking away at something with similar goals as this for a while. Mostly abandoned actually due to lack of time. Came across Sagui on twitter today and it looks like it could mostly fill my needs with some tweaking.

Differences being, no sass support, postcss, node.js testing only via tape, flowtyping, and a bunch of react packages, and some utility functions.

I was wondering if there's a good way to hook into sagui in order to modify/extend it. Would be nice for my company to be able to install one module and have everything setup. But it might also be better for the overall ecosystem to be able to install sagui and then install sagui plugins alongside it that modify how it works internally. (Other possibilities include different linting setups, additional/modified scripts, etc)

Let me know if you've thought about this kind of stuff already and maybe we can work something up.

**Comment by pirelenito** [comment]
Hi there! Thanks for reaching out.

I agree with you in that sense, I've built Sagui very much tailored to my "most common" needs and is very much an opinion on some of the stuff I think is more interesting to use nowadays.

Regarding extension and customization, that is indeed a hard problem, and Sagui's main reason for being on v6 already is my experimentation on trying to achieve that goal. I've tried going into a road of "presets" at one time to provide such "plugable" functionalities... the problem with that approach is that different pressets might step on each other toes, and you ended-up with build issues. I had real problems with this approach, and that is why that at this moment I decided to hide these "presets" as an implementation detail (as you can see in the source code).

There might be a solid possibility on creating "flavors" of this tool, which would cover different use cases and maybe reuse some core functionality and setup, but this is still an idea I'm developing. This sort of tools cold be the next iteration of something like Yeoman.

Regarding right now, my main goal is trying to achieve a solution that if you stay on it and use its conventions you are going to be happy. I do however provide some configurations parameters to tweak the internals, and escape hatches through [webpack](https://github.com/saguijs/sagui#webpack) and [karma](https://github.com/saguijs/sagui#karma) configurations and also ways of [disabling](https://github.com/saguijs/sagui#disabledloaders) the internal loaders, but these should be used as last resources.

Regarding some of your needs:

- Sagui does have postCSS, (which is used to provide CSS Modules and Autoprefixer), what would be your needs regarding that?
- We are adding support for flowtype soon 😀
- Adding support to testing with Tape and Node.js would just make the current solution bulkier (maybe the flavors idea I've mentioned earlier?)

Anyway, if you are interested we can talk more about and prototype some ideas.

Thanks again!

**Comment by SeeThruHead** [comment]
Yeah the flavours idea appeals to me because we could iterate towards a plugin architecture over time which is nice. I can take a look at forking sagui and changing the configuration to support what I have in my boilerplate. Problem is it's a little more specialized than what I would consider to be a typical 'flavour'.

Major changes for my 'flavour' would be a different testing pipeline.
Removal of sass
a few helper functions from https://github.com/influitive/reforge
adapting some boilerplate into sagui template from https://github.com/influitive/reforge-boilerplate

**Comment by xaviervia** [comment]
### SASS

I'm curious of if it would be OK for you having SASS available but not using it. Sagui does actually support plain CSS, and since I don't consider SASS necessary for most of my personal projects I've been using Sagui with plain CSS and postCSS, ignoring SASS. Just [adding files with `.css`](https://github.com/saguijs/sagui/blob/master/src/webpack/loaders/style.js#L62) extension works.

### Flowtype

tl;dr We are adding support soon

I also would like that added ASAP, we are working on that. Currently the salient issue is a compatibility problem with StandardJS: by switching the eslint parser to be `babel-eslint`, flowtypes are successfully ignored for linting, but the linter does not show any warnings for flow specific rules. 

I've been trying with https://github.com/gajus/eslint-plugin-flowtype . I added a [branch in the `babel-preset-sagui` repo](https://github.com/saguijs/babel-preset-sagui/blob/try-eslint-plugin-flowtype/test/fixture.js) were you can see an example of running the flowtype plugin against a [file with lint](https://github.com/saguijs/babel-preset-sagui/blob/try-eslint-plugin-flowtype/test/fixture.js), which for some reason shows no warnings. Cloning the babel-preset-sagui repo and running `npm test` runs the linter against that file.

For now we might go ahead with adding support without flowtypes linting, but any pointers on what the issue might be are greatly appreciated.

Let's continue this discussion in the flow type issue: https://github.com/saguijs/sagui/issues/33

### Tape.js

I don't fully understand why this would be a big issue @pirelenito. In many of my projects that don't use webpack I build everything with babel, including the tests, then run the tests by running a particular js entrypoint with node (something like `babel test.js `). The same could be done with sagui rather easily, that is, running a particular file as the entrypoint with `babel-node`, using the same babel presets. This works for me because the test library I'm using allows for [running the suite programmatically](https://github.com/xaviervia/washington#basic-example), so shell commands are not needed. As long as Tape or whatever is being used works without the need for using the CLI, it should be the same. 

> Note: running would have to be done with `babel-node` instead of webpack because otherwise it wouldn't be possible to get errors pointing to the correct line numbers.

As I see it, this could be a flag to set in the `sagui.config.js`, something like:

```javascript
test: {
  runner: {
    type: 'node',
    entrypoint: 'test'
  }
}
```

The default would be:

```javascript
test: {
  runner: {
    type: 'karma'
  }
}
```

Of course the assumption is that files that are to be tested with Node don't include CSS or any other type of webpack modules: if those are used, the runner should be karma.

I'd personally like this idea a lot because libraries such as tape favor small and focused unit testing, which is what I prefer, and also karma takes a lot of time to startup. 

> Maybe we should break up this discussion into it's own thread as well.

**Comment by pirelenito** [comment]
My main concern with extensibility is loosing one of the main things I was trying to solve when I started Sagui, remove most of the decision making that lingers in the Front-end community and have a quick setup that just works.

That being said, I've been thinking that we could make Sagui just a "platform" that provides core functionality (Webpack with core loaders) but leave the user to choose (which could be community contributed):

- Test runner
- Style
- JavaScript compiler
- Archetypes (pages, library and others)

And those would be external dependencies, so we would basically slim down Sagui (take the existing implementations of these solutions out) and leave the decision to the user. Maybe the CLI would ask on first install (but could have some nice defaults)?

I'll try prototyping a solution this week and see how it goes.

**Comment by SeeThruHead** [comment]
I like the idea of a less opinionated sagui-core. People could either use core directly if they need a lot of control. Or use 'sagui' (a layer on top of sagui) or 'insert cool name here' another layer on top of sagui that chooses different tooling.

**Comment by bichotll** [comment]
Hi, just adding my two cents here. It would be cool to have presets as we have with ESLint.
So every user/company can have a different configuration and create others.

**Comment by pirelenito** [comment]
We think it is not realistic to have extensibility like this in the current iteration of Sagui, maybe we can revisit this again in the future.

Thanks for the contributions.
